### PR TITLE
Add map_windows support to Go SDK

### PR DESCRIPTION
**Pull request by camphillips22**
Adds support to the Go SDK for the [map_windows][1] urn.

The existing type model does not allow `IntervalWindow` as a FullValue and the existing coders will throw away the window values when decoding the KV type [here][2] in `elideSingleElmFV`. This change updates coder construction to allow an IntervalWindow as a value coder.

Addresses #23106.

[1]: https://github.com/apache/beam/blob/master/model/pipeline/src/main/proto/org/apache/beam/model/pipeline/v1/beam_runner_api.proto#L296
[2]: https://github.com/camphillips22/beam/blob/8be4cdcaf65a1e53e3041ac3354e2e99c845e915/sdks/go/pkg/beam/core/runtime/exec/coder.go#L580-L580

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by camphillips22** [comment]
R: @lostluck

**Comment by camphillips22** [comment]
A note about testing: I was unable to write a batch pipeline that resulted in a window mapping operation. 

[Here](https://github.com/camphillips22/beam/compare/cam/mapwindows...camphillips22:beam:mapwindows-manual-tests?expand=1) is a link to the tests I wrote which did not result in a map_windows operation. It also contains the changes I made to the streaming wordcap example to make it happen on dataflow. I have not written a pipeline yet that works on Flink and uses map windows, but I wanted to go ahead get this change out for review.

File: sdks/go/pkg/beam/core/graph/coder/coder.go
Comment (lostluck) [inline sdks/go/pkg/beam/core/graph/coder/coder.go:180]: Note how nothing else in this `coder` package calls out it's coderness in their short names. I'd just call this `IntervalWindow` with IW as the short version.

The 'value' sufixes for WindowedValue and ParamWindowedValue indicate they're wrapping a value themselves, rather than being windows themselves, like the one we're adding here.

File: sdks/go/pkg/beam/core/runtime/graphx/coder.go
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/graphx/coder.go:251]: Per the proto "map_windows" spec, the input/output is "KV<[]byte,Window>" but as written here, this just returns "window".  It'd be better if we follow the normal structure, generalizing to handle interval windows better instead of special casing here.

File: sdks/go/pkg/beam/core/runtime/exec/window.go
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/exec/window.go:129]: As written here, it looks like we assume the output nonce is the same as the input window, which may not be the case in all runners. As the coder is written, and picked for a KV, it looks like the mapped element never gets coded.

**Comment by lostluck** [comment]
This is a very good start!

> A note about testing: I was unable to write a batch pipeline that resulted in a window mapping operation.
> 
> [Here](https://github.com/camphillips22/beam/compare/cam/mapwindows...camphillips22:beam:mapwindows-manual-tests?expand=1) is a link to the tests I wrote which did not result in a map_windows operation. It also contains the changes I made to the streaming wordcap example to make it happen on dataflow. I have not written a pipeline yet that works on Flink and uses map windows, but I wanted to go ahead get this change out for review.

I looked into the Dataflow's graph analyzer and it doesn't use map windows in batch, so that's not amazing for batch testing. We should at least manually validate in streaming mode. I can do this if you like.

We should still have unit tests for the MapWindows Node in the exec package. We should test to the spec (with the KV<nonce, Window> inputs and outputs), per the proto. eg.That we preserve the "nonce", and similar. https://github.com/apache/beam/blob/master/model/pipeline/src/main/proto/org/apache/beam/model/pipeline/v1/beam_runner_api.proto#L285 

I've got a few more specific comments in a moment or two.

**Review by lostluck — COMMENTED** [review]
And just a heads up, in order to unblock RunInference, I've got a change that makes slices use the beam Iterable coder by default, for better cross language support. It affects some overlapping code, so be aware of merge conflicts.

File: sdks/go/pkg/beam/core/graph/coder/coder.go
Comment (camphillips22) [reply sdks/go/pkg/beam/core/graph/coder/coder.go:180]: I had named it `IntervalWindow` to start, but there's a conflicting `IntervalWindow` value define [here](https://github.com/camphillips22/beam/blob/c83b42f7547a5fd57e2f9a90d7b613c512eb088c/sdks/go/pkg/beam/core/graph/coder/windows.go#L25-L25). Maybe just `IW` for const name with a comment that explicitly calls out that it refers to an `IntervalWindow`? Any other suggestions?

I can definitely change the value to "IW" though.

File: sdks/go/pkg/beam/core/runtime/graphx/coder.go
Comment (camphillips22) [reply sdks/go/pkg/beam/core/runtime/graphx/coder.go:251]: Whoops! I was wondering why the IW was in `elm.Elm` when I thought it should've been in `elm.Elm2`. Returning a KV coder here should also fix the `ProcessElement` to use `elm.Elm` for the nonce and `elm.Elm2` for the window (addressing your comment [here](https://github.com/apache/beam/pull/24307#discussion_r1029916481)).

Comment (camphillips22) [reply sdks/go/pkg/beam/core/runtime/graphx/coder.go]: > It'd be better if we follow the normal structure, generalizing to handle interval windows better instead of special casing here.

To clarify, are you suggesting that we need to pull this special case out of the KV coder case, or just to return an actual KV coder?

Comment (lostluck) [reply sdks/go/pkg/beam/core/runtime/graphx/coder.go]: The recursive lookups (to b.Coder) for the key and value types should be able to handle getting the nonce (AKA []byte), and the Interval Window coder.

Comment (lostluck) [reply sdks/go/pkg/beam/core/runtime/graphx/coder.go]: Basically, the Iterable/StateBackedIterable cases need special handling because technically CoGBK results are KV<k, Iter<V>> (or even KV<k, Iter<V1>, ... , Iter<Vn>>), but we avoid degenerate KVs with more than 2 components in the Go SDK, by having a dedicated CoGBK "kind" for coders.  

This lets us treat the data ingestion properly datasource.go, and be able to do the special handling to convert to the pull iterator functions `func(*V) bool` (and with any luck, once Go settles on an Iterator convention, support that style as well.)

Basically, if we had "real" KV support as a single KV element instead of splitting K and V parameters, we wouldn't need to go this far, but we're stuck with supporting it in the current API.

But this is a long way of saying, we don't need to special case things at this layer since there's nothing "tricky" about the element type technically, we just put Windows into their own box to avoid accidental mistakes (or generally nonsensical user mistakes like "processing" the internal windows metadata)

File: sdks/go/pkg/beam/core/runtime/graphx/coder.go
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/graphx/coder.go:253]: Ah didn't see this. Basically, we should only need to put the `case urnIntervalWindow: return coder.NewIntervalWindowCoder(), nil` into the outer switch (probably just above KV, since it's easier to read), and not explicitly handle it under the KV coder case.

Comment (camphillips22) [reply sdks/go/pkg/beam/core/runtime/graphx/coder.go:253]: Ah, so there is already a case in the switch for that [here](https://github.com/apache/beam/blob/master/sdks/go/pkg/beam/core/runtime/graphx/coder.go#L388-L393), and it talks about special handling for tests. I was hesitant to do anything with that before review. It also looks like there's other behavior that might depend on that being a `Window` coder instead of this new kind.

Comment (lostluck) [reply sdks/go/pkg/beam/core/runtime/graphx/coder.go:253]: As the person who put that in, I'd rather inconvenience the tests than the production code. Please feel free to change them up. (sorry for the churn).
That was before I was aware that there were uses for runner sending coders around like that, and we just wanted to validate the coders per the standard_coders.yaml tests.

Those uses are handled here usually: https://github.com/camphillips22/beam/tree/c83b42f7547a5fd57e2f9a90d7b613c512eb088c/sdks/go/test/regression/coders/fromyaml

File: sdks/go/pkg/beam/core/runtime/graphx/coder.go
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/graphx/coder.go:382]: Now that we have a coder* for it, we should ensure that the coder can be round trip serialized. Please add an Unmarshal(Marshal(*coder*)) round trip test (see graphx/coder_test.go)

Comment (lostluck) [reply sdks/go/pkg/beam/core/runtime/graphx/coder.go:382]: I think that's the only remaining thing for me to block on before merging.

Also, FYI, my foretold change: #24346 
At this point I'll handle the merge conflict instead.

Comment (camphillips22) [reply sdks/go/pkg/beam/core/runtime/graphx/coder.go:382]: Done!

**Comment by camphillips22** [comment]
@lostluck Anything left to do here other than resolve conflicts?

**Comment by lostluck** [comment]
I think just resolving the conflicts? I'll take another look after i finish lunch.

**Comment by lostluck** [comment]
Run Go PreCommit

**Comment by lostluck** [comment]
PRecommit failure was a flake, so just rerunning to be sure. Otherwise this LGTM.

**Review by lostluck — APPROVED** [review]

**Comment by lostluck** [comment]
Run Go PostCommit

**Comment by lostluck** [comment]
Run Go Flink ValidatesRunner

**Comment by lostluck** [comment]
Thank you very much for working through this! Hopefully now it's possible for someone to cobble together a PeriodicImpulse or similar and have functioning Slowly Changing Side Inputs.

**Comment by camphillips22** [comment]
Thanks for helping me get this merged! It will help tremendously with our pipelines.

**Pull request merged**
